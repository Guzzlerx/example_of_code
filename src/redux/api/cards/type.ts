import { IMyCardProductItem } from 'routes/Main/Cards/CardProducts/type';
import { ICardStatusSwitcherProps } from 'routes/Main/Cards/MyCards/MyCardDetails/MyCardDetailsTabGroup/Tabs/ManageTab/Parts/CardStatusSwitcher/type';
import { IMyCardItem } from 'routes/Main/Cards/MyCards/type';
import {
  CARD_NAMES_ENUM,
  CARD_STATUSES_ENUM,
  CARD_TYPES_ENUM,
  CURRENCY_CODES_ENUM,
  PAYMENT_SYSTEMS_ENUM,
} from 'constants/cards/cards';

export interface IMyCardsList {
  cards: IMyCardItem[];
}

export interface IMyCardDetails {
  creditLimit: string;
  principalDebt: string;
  terminationDate: string;
  accountId: string;
  accountNumber: string;
  balance: string;
  cardId: string;
  cardNumber: string;
  clientId: string;
  currencyCode: CURRENCY_CODES_ENUM;
  expirationDate: string;
  holderName: string;
  paymentSystem: PAYMENT_SYSTEMS_ENUM;
  productName: CARD_NAMES_ENUM;
  status: CARD_STATUSES_ENUM;
  transactionLimit: string;
}

export interface ICardProductsList {
  cardProducts: IMyCardProductItem[];
}

export type TUpdateCardStatusRequest = Omit<ICardStatusSwitcherProps, 'refetchHandler'>;
export interface INewCardRequest {
  idProduct?: string;
  holderName?: string;
  cardType?: CARD_TYPES_ENUM;
}
