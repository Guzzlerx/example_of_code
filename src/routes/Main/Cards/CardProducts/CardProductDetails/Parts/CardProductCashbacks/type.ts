import { CARD_NAMES_ENUM } from 'constants/cards/cards';

export interface ICardProductCashbacksProps {
  cardName: Exclude<
    CARD_NAMES_ENUM,
    CARD_NAMES_ENUM.SMART_PAYOUT | CARD_NAMES_ENUM.INFINITY_PREMIUM | CARD_NAMES_ENUM.PERSONAL_LOAN
  >;
}
