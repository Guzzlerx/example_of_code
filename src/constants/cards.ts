export enum CARD_TYPES_ENUM {
  DEBIT = 'DEBIT',
  CREDIT = 'CREDIT',
  DEPOSIT = 'DEPOSIT',
}

export enum CARD_STATUSES_ENUM {
  ACTIVE = 'ACTIVE',
  BLOCKED = 'BLOCKED',
  EXPIRED = 'EXPIRED',
}

export enum PAYMENT_SYSTEMS_ENUM {
  VISA = 'VISA',
  MASTERCARD = 'MASTERCARD',
}

export enum CARD_NAMES_ENUM {
  UNIVERSAL_CARD = 'universal_card',
  SMART_PAYOUT = 'smart_payout',
  PURCHASE_PLUS = 'purchase_plus',
  PERSONAL_LOAN = 'personal_loan',
  INFINITY_CLASSIC = 'infinity_classic',
  INFINITY_GOLD = 'infinity_gold',
  INFINITY_PREMIUM = 'infinity_premium',
}

export enum CURRENCY_CODES_ENUM {
  EUR = 'EUR',
  USD = 'USD',
  KZT = 'KZT',
  GBP = 'GBP',
  CHF = 'CHF',
  PLN = 'PLN',
  RUB = 'RUB',
}

export const MAX_CARDS_TO_SHOW = 3;

export enum CARD_SIZE_ENUM {
  LARGE = 'large',
  MEDIUM = 'medium',
  SMALL = 'small',
  EXTRA_SMALL = 'extraSmall',
}
