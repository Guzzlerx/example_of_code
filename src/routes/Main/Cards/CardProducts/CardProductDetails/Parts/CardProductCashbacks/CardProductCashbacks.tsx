import { useTranslation } from 'react-i18next';
import { ATitle } from 'components/ATitle/ATitle';
import { ReactComponent as AdvantagesGraph } from 'assets/image/cardDetailsInfo/advantage_graph.svg';
import {
  StyledAdvantagesContainer,
  StyledAdvantagesWrapper,
  StyledText,
} from '../CardProductAdvantages/CardProductAdvantages.styles';
import { ICardProductCashbacksProps } from './type';

export const CardProductCashbacks = ({ cardName }: ICardProductCashbacksProps) => {
  const { t } = useTranslation('cards');

  const translations = {
    firstCashbackTitle: t(`card_details.advantages.${cardName}.cashbacks.first_cashback.title`),
    firstCashbackDescription: t(
      `card_details.advantages.${cardName}.cashbacks.first_cashback.description`,
    ),
    secondCashbackTitle: t(`card_details.advantages.${cardName}.cashbacks.second_cashback.title`),
    secondCashbackDescription: t(
      `card_details.advantages.${cardName}.cashbacks.second_cashback.description`,
    ),
    thirdCashbackTitle: t(`card_details.advantages.${cardName}.cashbacks.third_cashback.title`),
    thirdCashbackDescription: t(
      `card_details.advantages.${cardName}.cashbacks.third_cashback.description`,
    ),
  };

  const cardCashbacksData = [
    {
      title: translations.firstCashbackTitle,
      description: translations.firstCashbackDescription,
    },
    {
      title: translations.secondCashbackTitle,
      description: translations.secondCashbackDescription,
    },
    {
      title: translations.thirdCashbackTitle,
      description: translations.thirdCashbackDescription,
    },
  ];

  return (
    <StyledAdvantagesContainer>
      <AdvantagesGraph />
      {cardCashbacksData.map(({ title, description }, index) => (
        <StyledAdvantagesWrapper width={196} marginTop={35} key={index}>
          <ATitle size={20} lineHeight={28} weight={400}>
            {title}
          </ATitle>
          <StyledText>{description}</StyledText>
        </StyledAdvantagesWrapper>
      ))}
    </StyledAdvantagesContainer>
  );
};
