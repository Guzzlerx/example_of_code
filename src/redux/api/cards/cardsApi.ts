import { apiSlice } from 'redux/api/apiSlice';
import { ICardStatusSwitcherResponse } from 'routes/Main/Cards/MyCards/MyCardDetails/MyCardDetailsTabGroup/Tabs/ManageTab/Parts/CardStatusSwitcher/type';
import { CARD_TYPES_ENUM } from 'constants/cards/cards';
import {
  ICardProductsList,
  IMyCardDetails,
  IMyCardsList,
  INewCardRequest,
  TUpdateCardStatusRequest,
} from './type';

export const cardsApi = apiSlice.injectEndpoints({
  endpoints: builder => ({
    getCreditCards: builder.query<IMyCardsList, void>({
      query: () => '/credit-cards',
    }),
    getDebitCards: builder.query<IMyCardsList, void>({
      query: () => '/deposit-cards',
    }),
    getCardDetails: builder.query<IMyCardDetails, { id: string; type: CARD_TYPES_ENUM }>({
      query: ({ id, type }) => `/${type.toLowerCase()}-cards/${id}`,
    }),
    getCardProducts: builder.query<ICardProductsList, void>({
      query: () => '/cards-products',
    }),
    updateCardStatus: builder.mutation<ICardStatusSwitcherResponse, TUpdateCardStatusRequest>({
      query: ({ cardId, currentStatus, cardType }) => ({
        url: `${cardType.toLocaleLowerCase()}-cards/${cardId}`,
        method: 'PATCH',
        body: currentStatus,
      }),
    }),
    getNewCard: builder.mutation<void, INewCardRequest>({
      query: body => ({
        url: '/deposit-cards-orders/new',
        method: 'POST',
        body: body,
      }),
    }),
  }),
});

export const {
  useGetCreditCardsQuery,
  useGetDebitCardsQuery,
  useGetCardProductsQuery,
  useGetCardDetailsQuery,
  useUpdateCardStatusMutation,
  useGetNewCardMutation,
} = cardsApi;
