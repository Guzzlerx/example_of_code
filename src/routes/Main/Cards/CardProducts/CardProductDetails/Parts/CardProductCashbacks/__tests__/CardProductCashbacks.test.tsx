import { RenderResult, render, screen } from '@testing-library/react';
import { I18nextProvider } from 'react-i18next';
import { CARD_NAMES_ENUM } from 'constants/cards/cards';
import i18n from 'utils/languages/i18nForTests';
import { CardProductCashbacks } from '../CardProductCashbacks';

describe('CardProductCashbacks', () => {
  let renderResult: RenderResult;

  beforeEach(() => {
    renderResult = render(
      <I18nextProvider i18n={i18n}>
        <CardProductCashbacks cardName={CARD_NAMES_ENUM.UNIVERSAL_CARD} />
      </I18nextProvider>,
    );
  });

  it('should render correctly', () => {
    expect(renderResult.container).toBeInTheDocument();
  });

  it('snapshot', () => {
    expect(renderResult.container).toMatchSnapshot();
  });

  it('render correctly with "infinity classic" card as prop', () => {
    renderResult.rerender(<CardProductCashbacks cardName={CARD_NAMES_ENUM.INFINITY_CLASSIC} />);

    const elements = screen.getAllByText(/cashback/i);

    expect(elements[0]).toBeInTheDocument();
    expect(elements[1]).toBeInTheDocument();
    expect(elements[2]).toBeInTheDocument();

    expect(
      screen.getByText(/On your first GPB 4,000 eligible purchases with/i),
    ).toBeInTheDocument();
    expect(screen.getByText(/On your first GPB 4,000 eligible purchases on/i)).toBeInTheDocument();
    expect(screen.getByText(/On 5 categories of chosen products/i)).toBeInTheDocument();
  });

  it('render correctly with "infinity gold" card as prop', () => {
    renderResult.rerender(<CardProductCashbacks cardName={CARD_NAMES_ENUM.INFINITY_GOLD} />);

    const elements = screen.getAllByText(/cashback/i);

    expect(elements[0]).toBeInTheDocument();
    expect(elements[1]).toBeInTheDocument();
    expect(elements[2]).toBeInTheDocument();

    expect(screen.getByText(/On 5 categories of chosen products/i)).toBeInTheDocument();
    expect(
      screen.getByText(/On your first GPB 4,000 eligible purchases with/i),
    ).toBeInTheDocument();
    expect(screen.getByText(/On all other eligible purchases/i)).toBeInTheDocument();
  });

  it('render correctly with "purchase plus" card as prop', () => {
    renderResult.rerender(<CardProductCashbacks cardName={CARD_NAMES_ENUM.PURCHASE_PLUS} />);

    const elements = screen.getAllByText(/cashback/i);

    expect(elements[0]).toBeInTheDocument();
    expect(elements[1]).toBeInTheDocument();
    expect(elements[2]).toBeInTheDocument();

    expect(
      screen.getByText(/On your first GPB 8,000 eligible purchases with a retailer you choose/i),
    ).toBeInTheDocument();
    expect(
      screen.getByText(/On your first GPB 8,000 eligible purchases on your bills/i),
    ).toBeInTheDocument();
    expect(screen.getByText(/On all other purchases at all amounts/i)).toBeInTheDocument();
  });
});
