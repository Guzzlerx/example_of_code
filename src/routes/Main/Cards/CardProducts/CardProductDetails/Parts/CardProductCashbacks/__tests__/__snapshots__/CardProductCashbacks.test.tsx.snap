// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardProductCashbacks snapshot 1`] = `
<div>
  <div
    class="sc-dkrFOg dbAEKA"
  >
    <svg>
      advantage_graph.svg
    </svg>
    <div
      class="sc-hLBbgP jkPHwm"
      width="196"
    >
      <h2
        class="sc-bcXHqe iHLRVe"
        size="20"
      >
        8% cashback
      </h2>
      <p
        class="sc-eDvSVe OfAwA"
      >
        On your first GPB 2,000 eligible purchases with two retailers you choose
      </p>
    </div>
    <div
      class="sc-hLBbgP jkPHwm"
      width="196"
    >
      <h2
        class="sc-bcXHqe iHLRVe"
        size="20"
      >
        3% cashback
      </h2>
      <p
        class="sc-eDvSVe OfAwA"
      >
        On your first GPB 2,000 eligible purchases on your bills
      </p>
    </div>
    <div
      class="sc-hLBbgP jkPHwm"
      width="196"
    >
      <h2
        class="sc-bcXHqe iHLRVe"
        size="20"
      >
        2% cashback
      </h2>
      <p
        class="sc-eDvSVe OfAwA"
      >
        On all other eligible purchases
      </p>
    </div>
  </div>
</div>
`;
